### PR TITLE
Add TargetKind reconciler

### PR DESCRIPTION
**Pull request by ksputo**
**Description**

Changes proposed in this pull request:

- Added TargetKind reconciliation process
- Added logic for generic injection of binding labels
- ...

**Related issue(s)**
#9402

### CODE REVIEW

File: components/binding/internal/controller/targetkind_reconciler.go [hunk 1]
```diff
@@ -11,25 +13,56 @@ import (
...
 	}
 }
 
 func (r *TargetKindReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	_ = context.Background()
+	ctx := context.Background()
+	var targetKind bindingsv1alpha1.TargetKind
+	err := r.client.Get(ctx, req.NamespacedName, &targetKind)
+	if err != nil {
+		r.log.Warnf("TargetKind %s not found during reconcile process: %s", req.NamespacedName, err)
```
Comment (polskikiel) [inline components/binding/internal/controller/targetkind_reconciler.go:42]: why we don't return error here? Are you sure it's not found error?

File: components/binding/internal/storage/storage.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	return nil
+}
+
+// Get returns KindManager for given Kind
+func (s *KindStorage) Get(kind Kind) (*ResourceData, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	concreteResourceData, exists := s.registered[kind]
+	if !exists {
+		return &ResourceData{}, errors.New(fmt.Sprintf("TargetKind %s was not found", kind))
```
Comment (polskikiel) [inline components/binding/internal/storage/storage.go:73]: it's better to use `Errorf` - btw. can we return here a typed error?

File: components/binding/internal/worker/targetkind.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	log.Infof("TargetKind %q registered", targetKind.Name)
+	return targetKind, nil
+}
+
+func (w *TargetKindWorker) reconcileUponAddUpdate(targetKind v1alpha1.TargetKind) error {
+	return w.storage.Register(targetKind)
+}
+
+func (w *TargetKindWorker) reconcileUponDelete(targetKind v1alpha1.TargetKind) error {
+
```
Comment (polskikiel) [inline components/binding/internal/worker/targetkind.go:62]: unnecessary whitespace

**Review by polskikiel — CHANGES_REQUESTED** [review]

**Comment by jasiu001** [comment]
make `go fmt` for all files

File: components/binding/internal/worker/targetkind.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+	targetKind.Status.Registered = true
+	log.Infof("TargetKind %q registered", targetKind.Name)
+	return targetKind, nil
+}
+
+func (w *TargetKindWorker) reconcileUponAddUpdate(targetKind v1alpha1.TargetKind) error {
+	return w.storage.Register(targetKind)
+}
+
+func (w *TargetKindWorker) reconcileUponDelete(targetKind v1alpha1.TargetKind) error {
```
Comment (jasiu001) [inline components/binding/internal/worker/targetkind.go:61]: unused method

File: components/binding/internal/worker/targetkind.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+	err := w.reconcileUponAddUpdate(*targetKind)
+	if err != nil {
+		return &v1alpha1.TargetKind{}, errors.New(fmt.Sprintf("while registering TargetKind %q", targetKind.Spec.DisplayName))
+	}
+	targetKind.Status.Registered = true
+	log.Infof("TargetKind %q registered", targetKind.Name)
+	return targetKind, nil
+}
+
+func (w *TargetKindWorker) reconcileUponAddUpdate(targetKind v1alpha1.TargetKind) error {
```
Comment (jasiu001) [inline components/binding/internal/worker/targetkind.go:57]: Must this be a separate method?

File: components/binding/internal/worker/targetkind.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+func NewTargetKindWorker(storage TargetKindStorage, dynamicClient dynamic.Interface) *TargetKindWorker {
+	return &TargetKindWorker{storage: storage,
+	dynamicClient: dynamicClient,
+	}
+}
+
+func (w *TargetKindWorker) Process(targetKind *v1alpha1.TargetKind, log log.FieldLogger) (*v1alpha1.TargetKind, error) {
+	log.Info("start TargetKind process")
+	// create client for kind and save to storage
```
Comment (jasiu001) [inline components/binding/internal/worker/targetkind.go:32]: I think the comment is out of date :)

File: components/binding/internal/worker/targetkind.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+	if targetKind.Status.Registered {
+	_, err := w.storage.Get(storage.Kind(targetKind.Name))
```
Comment (jasiu001) [inline components/binding/internal/worker/targetkind.go]: IMO is better to make a field `targetKind.Name` as `Kind` type inside `/pkg/apis/v1alpha1/targetkind_types.go` then you don't have to map all the time

File: components/binding/internal/worker/targetkind.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+	dynamicClient dynamic.Interface
+}
+
+func NewTargetKindWorker(storage TargetKindStorage, dynamicClient dynamic.Interface) *TargetKindWorker {
+	return &TargetKindWorker{storage: storage,
+	dynamicClient: dynamicClient,
+	}
+}
+
+func (w *TargetKindWorker) Process(targetKind *v1alpha1.TargetKind, log log.FieldLogger) (*v1alpha1.TargetKind, error) {
```
Comment (jasiu001) [inline components/binding/internal/worker/targetkind.go:30]: you use the same block of code twice, maybe easier will be:
```go
	if targetKind.Status.Registered {
		if w.storage.Is(targetKind.Name) {
			log.Infof("TargetKind %q was already registered", targetKind.Name)
			return targetKind, nil
		}
	}
	err := w.reconcileUponAddUpdate(*targetKind)
	if err != nil {
		return &v1alpha1.TargetKind{}, errors.New(fmt.Sprintf("while registering TargetKind %q", targetKind.Spec.DisplayName))
	}
	targetKind.Status.Registered = true
	log.Infof("TargetKind %q registered", targetKind.Name)
	return targetKind, nil
```

File: components/binding/internal/target/handler.go [hunk 9]
```diff
@@ -1,134 +1,223 @@
...
-	"github.com/google/uuid"
 	"github.com/pkg/errors"
-	v1 "k8s.io/api/apps/v1"
-	"sigs.k8s.io/controller-runtime/pkg/client"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/dynamic"
 )
 
 // TODO: change Handler methods when TargetKind logic will be ready
```
Comment (jasiu001) [inline components/binding/internal/target/handler.go:17]: remove TODO, you made this already :)

File: components/binding/internal/target/handler.go [hunk 10]
```diff
@@ -1,134 +1,223 @@
...
 
 	return nil
 }
 
-func (h *Handler) getDeployment(ctx context.Context, b *v1alpha1.Binding) (*v1.Deployment, error) {
-	var deployment v1.Deployment
+func (h *Handler) getResource(b *v1alpha1.Binding, data *storage.ResourceData) (*unstructured.Unstructured, error) {
+	resource, err := h.client.Resource(data.Schema).Namespace(b.Namespace).Get(b.Spec.Target.Name, metav1.GetOptions{})
+	if err != nil {
+		return nil, errors.Wrap(err, "while getting resource")
```
Comment (jasiu001) [inline components/binding/internal/target/handler.go:146]: maybe as a temporary?

File: components/binding/internal/target/handler.go [hunk 11]
```diff
@@ -1,134 +1,223 @@
...
+	return fmt.Sprintf("%s-%s", v1alpha1.BindingLabelKey, b.Name)
 }
 
-func (h *Handler) updateDeployment(ctx context.Context, deployment *v1.Deployment) error {
-	err := h.client.Update(ctx, deployment)
+func (h *Handler) getResourceLabels(res *unstructured.Unstructured, labelFields []string) (map[string]string, error) {
+	labels, err := h.findOrCreateLabelsField(res, labelFields)
 	if err != nil {
-		return bindErr.AsTemporaryError(err, "while updating target deployment by client")
+		return make(map[string]string, 0), err
```
Comment (jasiu001) [inline components/binding/internal/target/handler.go:166]: wrap all errors if has sense

File: components/binding/internal/target/handler.go [hunk 12]
```diff
@@ -1,134 +1,223 @@
...
+	}
+
+	result := make(map[string]string, 0)
+	for key, value := range labels {
+		strKey := fmt.Sprintf("%v", key)
```
Comment (jasiu001) [inline components/binding/internal/target/handler.go:171]: what is the reason to make this `fmt.Sprintf("%v", key)`?

File: components/binding/internal/target/handler.go [hunk 13]
```diff
@@ -1,134 +1,223 @@
...
+	if err != nil {
+		return errors.Wrapf(err, "while getting resource for Binding %s", b.Name)
 	}
 
-	for key, _ := range deployment.Spec.Template.ObjectMeta.Labels {
+	resourceLabels, err := h.getResourceLabels(resource, resourceData.LabelFields)
+	if err != nil {
+		return errors.Wrapf(err, "while getting injected labels for Binding %s", b.Name)
+	}
+	for key, _ := range resourceLabels {
```
Comment (jasiu001) [inline components/binding/internal/target/handler.go:89]: you can ommit:
```go
for key := range resourceLabels {
...
```

File: components/binding/internal/target/handler.go [hunk 14]
```diff
@@ -1,134 +1,223 @@
...
 		if key == h.labelKey(b) {
-			deployment.Spec.Template.ObjectMeta.Labels[h.labelKey(b)] = uuid.New().String()
```
Comment (jasiu001) [inline components/binding/internal/target/handler.go:73]: I don't understand this part: you fetch all labels, then iterate for keys, and if a label exists you `ensureLabelsAreApplied` ?

File: components/binding/internal/target/handler.go [hunk 15]
```diff
@@ -1,134 +1,223 @@
...
 
-	newLabels := make(map[string]string, 0)
-	for key, value := range deployment.Spec.Template.ObjectMeta.Labels {
-		if key == h.labelKey(b) {
+	for key, value := range existingLabels {
+		if key != h.labelKey(b) {
 			continue
 		}
-		newLabels[key] = value
+		labelsToDelete[key] = value
```
Comment (jasiu001) [inline components/binding/internal/target/handler.go:129]: IMO it has no sense, the only label in `labelsToDelete` for this concrete Binding will be: `h.labelKey(b)`
and value is not necessary, if I am correct you not use value inside `ensureLabelsAreDeleted()` method

**Review by jasiu001 — CHANGES_REQUESTED** [review]

**Review by polskikiel — APPROVED** [review]
lgtm

File: components/binding/pkg/apis/v1alpha1/binding_types.go [hunk 16]
```diff
@@ -36,7 +36,7 @@ type Source struct {
 }
 
 type Target struct {
-	Kind string `json:"kind"`
+	Kind Kind `json:"kind"`
```
Comment (jasiu001) [inline components/binding/pkg/apis/v1alpha1/binding_types.go:39]: leave string here

File: components/binding/pkg/apis/v1alpha1/targetkind_types.go [hunk 17]
```diff
@@ -4,22 +4,25 @@ import (
...
+	DisplayName string    `json:"displayName"`
+	Resource    Resource `json:"resource"`
+	LabelsPath  string    `json:"labelsPath"`
+}
 
-	// Foo is an example field of TargetKind. Edit TargetKind_types.go to remove/update
-	Foo string `json:"foo,omitempty"`
+type Resource struct {
+	Group   string `json:"group"`
+	Kind    Kind `json:"kind"`
```
Comment (jasiu001) [inline components/binding/pkg/apis/v1alpha1/targetkind_types.go:19]: make `go fmt` for all files

File: components/binding/internal/worker/targetkind.go [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+
+	if targetKind.Status.Registered {
+		if w.storage.Exist(targetKind.Spec.Resource.Kind) {
+			log.Infof("TargetKind %q was already registered", targetKind.Name)
+			return targetKind, nil
+		}
+	}
+	err := w.storage.Register(*targetKind)
+	if err != nil {
+		return &v1alpha1.TargetKind{}, errors.New(fmt.Sprintf("while registering TargetKind %q", targetKind.Spec.DisplayName))
```
Comment (jasiu001) [inline components/binding/internal/worker/targetkind.go:42]: return `targetKind` not `&v1alpha1.TargetKind{}`

File: components/binding/internal/storage/storage.go [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+		return &ResourceData{}, fmt.Errorf("TargetKind %s was not found", kind)
+	}
+	return concreteResourceData, nil
+}
+
+func (s *KindStorage) Exist(kind v1alpha1.Kind) bool {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	_, exists := s.registered[kind]
+	return exists
```
Comment (jasiu001) [inline components/binding/internal/storage/storage.go:71]: check if `LabelsPath` is the same

File: components/binding/internal/target/handler.go [hunk 20]
```diff
@@ -1,134 +1,222 @@
...
-		deployment.Spec.Template.ObjectMeta.Labels = make(map[string]string, 0)
+	resource, err := h.getResource(b, resourceData)
+	if err != nil {
+		return errors.Wrapf(err, "while getting resource for Binding %s", b.Name)
 	}
-	deployment.Spec.Template.ObjectMeta.Labels[h.labelKey(b)] = uuid.New().String()
-
-	if err := h.updateDeployment(ctx, deployment); err != nil {
-		return err
+	logrus.Infof("Got resource: %+v", resource)
```
Comment (jasiu001) [inline components/binding/internal/target/handler.go:40]: remove single logrus

**Review by jasiu001 — CHANGES_REQUESTED** [review]

**Review by jasiu001 — APPROVED** [review]

**Pull request merged**